File: decision_mesh/edge.py
from __future__ import annotations

from typing import TYPE_CHECKING

import numpy as np

from ._helpers import _r, _rn
from .face import Face
from .vertex import Vertex

if TYPE_CHECKING:
    from .mesh import DecisionMesh


class Edge:
    """
    Edges should have the following information:

    vertices at each end
    a test (test_vertex) to check which side of the edge a point lies on
    the faces attached to the edge (+ and/or -)
    activity of the edge
    """
    _seq = 0

    def __init__(self, mesh: DecisionMesh, vertex0: Vertex, vertex1: Vertex, active: bool):
        self._id = Edge._seq; Edge._seq += 1
        self.vertex0 = vertex0
        self.vertex1 = vertex1
        self.mesh = mesh
        self.faces = dict({'+': None, '-': None})
        self.disqualifying = set()
        self.opposing_vertices = dict({'+': None, '-': None})

        dx = vertex1.x - vertex0.x
        dy = vertex1.y - vertex0.y

        # unit normal — stored as numpy array for bulk operations, and as scalars for hot paths
        n0 = -dy
        n1 = dx
        self.length = (n0 * n0 + n1 * n1) ** 0.5
        inv_len = 1.0 / self.length if self.length > 0 else 0.0
        n0 *= inv_len
        n1 *= inv_len
        self.normal = np.array([n0, n1], dtype=float)
        self._n0 = n0
        self._n1 = n1

        # signed line equation: normal * p = intercept
        self.intercept = n0 * vertex0.x + n1 * vertex0.y
        self.midpoint = None
        self.sub_edges = {'0': None, '1': None, '+': None, '-': None}
        self.active = False
        if active:
            self.activate()

    @property
    def sid(self) -> str:
        return f"E{self._id:03d}"

    def __repr__(self):
        v0 = getattr(self.vertex0, "sid", "V?")
        v1 = getattr(self.vertex1, "sid", "V?")
        # show which faces are attached by path (not full repr to avoid recursion)
        fp = self.faces.get('+'); fm = self.faces.get('-')
        fp_tag = f"{getattr(fp, 'sid', None)}:{getattr(fp, 'path', None)}" if fp else None
        fm_tag = f"{getattr(fm, 'sid', None)}:{getattr(fm, 'path', None)}" if fm else None
        has_mid = hasattr(self, "midpoint")
        subs = getattr(self, "sub_edges", None)
        sub_ready = [k for k in ('0', '1', '+', '-') if isinstance(subs, dict) and subs.get(k) is not None]

        return (f"<{self.sid} {v0}->{v1} active={self.active} "
                f"n={_rn(self.normal)} b={_r(self.intercept)} "
                f"faces(+={fp_tag},-={fm_tag}) mid={has_mid} sub={sub_ready}>")

    __str__ = __repr__

    def activate(self):
        if self.active:
            return
        self.active = True
        self.mesh.active_edges.add(self)
        self.add_midpoint()
        self.vertex0.add_edge(self)
        self.vertex1.add_edge(self)
        self.sub_edges = {'0': Edge(self.mesh, self.vertex0, self.midpoint, False), '1': Edge(self.mesh, self.midpoint, self.vertex1, False), '+': None, '-': None}

    def add_midpoint(self):
        if self.midpoint is not None:
            return
        self.midpoint = Vertex(self.mesh, (self.vertex0.x + self.vertex1.x)/2, (self.vertex0.y + self.vertex1.y)/2, parent_edge=self)

    def split(self):
        for sign in ('+', '-'):
            while self.faces.get(sign) is not None:
                self.faces[sign].split(self)

        self.sub_edges['0'].activate()
        self.sub_edges['1'].activate()
        self.deactivate()

    def deactivate(self):
        if not self.active:
            return
        self.active = False
        self.mesh.active_edges.discard(self)
        self.vertex0.remove_edge(self)
        self.vertex1.remove_edge(self)

    def test_vertex(self, v: Vertex) -> float:
        return self._n0 * v.x + self._n1 * v.y - self.intercept

    def other_vertex(self, v: Vertex) -> Vertex | None:
        if v is self.vertex0: return self.vertex1
        if v is self.vertex1: return self.vertex0
        return None

    def add_face(self, face: Face):
        """
        Attach a face to this edge and precompute its subdivision.

        Uses index-based point splitting: iterates only the parent face's
        coords_indices (not all N data points) and routes each point to
        the correct child face via the chord's half-plane test.
        """
        idx = face.edges.index(self)
        opposing_vertex = face.vertices[idx]
        if self.test_vertex(opposing_vertex) > 0:
            face_type = '+'
        else:
            face_type = '-'

        self.faces[face_type] = face
        self.opposing_vertices[face_type] = opposing_vertex

        self.sub_edges[face_type] = Edge(self.mesh, self.midpoint, opposing_vertex, False)
        idx = face.vertices.index(self.vertex1)
        edge0 = face.edges[idx]
        idx = face.vertices.index(self.vertex0)
        edge1 = face.edges[idx]

        # Split parent face's points by chord half-plane
        # Iterate only parent's coords_indices — O(points_in_parent) not O(N_total)
        chord = self.sub_edges[face_type]
        cn0 = chord._n0
        cn1 = chord._n1
        cint = chord.intercept

        X = self.mesh.X
        indices0 = []
        indices1 = []

        for i in face.coords_indices:
            val = cn0 * X[i, 0] + cn1 * X[i, 1]
            on_plus = val >= cint
            if face_type == '+':
                if on_plus:
                    indices0.append(i)
                else:
                    indices1.append(i)
            else:
                if not on_plus:
                    indices0.append(i)
                else:
                    indices1.append(i)

        if face_type == '+':
            path0 = face.path + '+'
            path1 = face.path + '-'
        else:
            path0 = face.path + '-'
            path1 = face.path + '+'

        # Create child faces with skip_coords=True, then use fast index path
        face0 = Face(self.mesh, edge0, self.sub_edges['0'], self.sub_edges[face_type], None, False, path0, skip_coords=True)
        face0.update_coords_from_indices(indices0)

        face1 = Face(self.mesh, edge1, self.sub_edges[face_type], self.sub_edges['1'], None, False, path1, skip_coords=True)
        face1.update_coords_from_indices(indices1)

        if face_type == '+':
            face.add_sub_division(self, {'e': self.sub_edges[face_type], '+': face0, '-': face1})
        else:
            face.add_sub_division(self, {'e': self.sub_edges[face_type], '+': face1, '-': face0})

        if max(face0.aspect_ratio(), face1.aspect_ratio()) >= self.mesh.max_aspect_ratio:
            self.disqualifying.add(face)
            if not self.midpoint.disqualified:
                self.midpoint.disqualified = True
                self.mesh.loss_heap.pop(self.midpoint, None)

        self.midpoint.update_info()

    def remove_face(self, face):
        if self.faces['+'] is face:
            self.faces['+'] = None
        if self.faces['-'] is face:
            self.faces['-'] = None

        self.disqualifying.discard(face)
        if len(self.disqualifying) == 0:
            self.midpoint.disqualified = False

File: cpp/edge.cpp
#include "edge.h"
#include "mesh.h"
#include "vertex.h"
#include "face.h"
#include <cmath>
#include <algorithm>

int Edge::_seq = 0;

Edge::Edge(DecisionMesh* mesh, Vertex* v0, Vertex* v1, bool active_flag)
    : _id(Edge::_seq++), vertex0(v0), vertex1(v1), mesh(mesh)
{
    double dx = v1->x - v0->x;
    double dy = v1->y - v0->y;
    double n0 = -dy, n1 = dx;
    length = std::hypot(n0, n1);
    if (length > 0) {
        normal[0] = n0 / length;
        normal[1] = n1 / length;
    }
    intercept = normal[0] * v0->x + normal[1] * v0->y;

    if (active_flag) {
        activate();
    }
}

std::string Edge::sid() const {
    char buf[16];
    snprintf(buf, sizeof(buf), "E%03d", _id);
    return buf;
}

void Edge::activate() {
    if (active) return;
    active = true;
    mesh->active_edges.insert(this);
    add_midpoint();
    vertex0->add_edge(this);
    vertex1->add_edge(this);

    // Create half sub-edges (inactive)
    sub_edges[0] = mesh->make_edge(vertex0, midpoint, false);  // '0'
    sub_edges[1] = mesh->make_edge(midpoint, vertex1, false);  // '1'
}

void Edge::add_midpoint() {
    if (midpoint) return;
    double mx = (vertex0->x + vertex1->x) / 2.0;
    double my = (vertex0->y + vertex1->y) / 2.0;
    midpoint = mesh->make_vertex(mx, my, false, this, true);
}

void Edge::split() {
    // Split faces on both sides
    for (int s = 0; s < 2; ++s) {
        while (faces[s] != nullptr) {
            faces[s]->split(this);
        }
    }

    sub_edges[0]->activate();  // '0'
    sub_edges[1]->activate();  // '1'
    deactivate();
}

void Edge::deactivate() {
    if (!active) return;
    active = false;
    mesh->active_edges.erase(this);
    vertex0->remove_edge(this);
    vertex1->remove_edge(this);
}

double Edge::test_vertex(Vertex* v) const {
    return normal[0] * v->x + normal[1] * v->y - intercept;
}

Vertex* Edge::other_vertex(Vertex* v) const {
    if (v == vertex0) return vertex1;
    if (v == vertex1) return vertex0;
    return nullptr;
}

void Edge::add_face(Face* face) {
    // Find opposing vertex
    int idx = -1;
    for (int i = 0; i < 3; ++i) {
        if (face->edges[i] == this) { idx = i; break; }
    }
    Vertex* opposing = face->vertices[idx];
    int si = (test_vertex(opposing) > 0) ? 0 : 1;  // 0='+', 1='-'
    char face_type = (si == 0) ? '+' : '-';

    faces[si] = face;
    opposing_vertices[si] = opposing;

    // Create internal chord sub-edge
    int sub_idx = si + 2;  // '+' -> 2, '-' -> 3
    sub_edges[sub_idx] = mesh->make_edge(midpoint, opposing, false);

    // Find edges opposite to vertex0 and vertex1
    int idx_v1 = -1, idx_v0 = -1;
    for (int i = 0; i < 3; ++i) {
        if (face->vertices[i] == vertex1) idx_v1 = i;
        if (face->vertices[i] == vertex0) idx_v0 = i;
    }
    Edge* edge0 = face->edges[idx_v1];
    Edge* edge1 = face->edges[idx_v0];

    // Build mask split using the chord's half-plane
    Edge* chord = sub_edges[sub_idx];
    std::vector<bool> chord_mask(mesh->n_points, false);
    for (int i = 0; i < mesh->n_points; ++i) {
        double val = chord->normal[0] * mesh->get_x(i, 0) +
                     chord->normal[1] * mesh->get_x(i, 1);
        chord_mask[i] = (val >= chord->intercept);
    }

    std::vector<bool> mask0(mesh->n_points), mask1(mesh->n_points);
    std::string path0, path1;

    if (face_type == '+') {
        for (int i = 0; i < mesh->n_points; ++i) {
            mask0[i] = chord_mask[i] && face->mask[i];
            mask1[i] = !chord_mask[i] && face->mask[i];
        }
        path0 = face->path + '+';
        path1 = face->path + '-';
    } else {
        for (int i = 0; i < mesh->n_points; ++i) {
            mask0[i] = !chord_mask[i] && face->mask[i];
            mask1[i] = chord_mask[i] && face->mask[i];
        }
        path0 = face->path + '-';
        path1 = face->path + '+';
    }

    Face* face0 = mesh->make_face(edge0, sub_edges[0], sub_edges[sub_idx], mask0, false, path0);
    Face* face1 = mesh->make_face(edge1, sub_edges[sub_idx], sub_edges[1], mask1, false, path1);

    SubDivision sd;
    sd.e = chord;
    if (face_type == '+') {
        sd.plus_face = face0;
        sd.minus_face = face1;
    } else {
        sd.plus_face = face1;
        sd.minus_face = face0;
    }
    face->add_sub_division(this, sd);

    // Check aspect ratio
    double ar = std::max(face0->aspect_ratio(), face1->aspect_ratio());
    if (ar >= mesh->max_aspect_ratio) {
        disqualifying.insert(face);
        if (!midpoint->disqualified) {
            midpoint->disqualified = true;
            mesh->heap_pop(midpoint);
        }
    }

    midpoint->update_info();
}

void Edge::remove_face(Face* face) {
    if (faces[0] == face) faces[0] = nullptr;
    if (faces[1] == face) faces[1] = nullptr;

    disqualifying.erase(face);
    if (disqualifying.empty()) {
        midpoint->disqualified = false;
    }
}
